### PR TITLE
Update and improve ConfigMap resource lifecycle test

### PR DESCRIPTION
**Pull request by BobyMCbobs**
- [x] [mock test ticket](https://github.com/kubernetes/kubernetes/issues/87120)
- [x] [PR](https://github.com/kubernetes/kubernetes/pull/87212)
- [ ] [promotion](https://github.com/kubernetes/kubernetes/pull/88034)
  - [ ] [update to improve](#)

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Adds more checks throughout the test:
- created
- modified
- deleted

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/pull/88034#discussion_r384103741

**Special notes for your reviewer**:
Should fix flakiness

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
NONE
```

/sig testing

### CODE REVIEW

File: test/e2e/common/configmap.go [hunk 1]
```diff
@@ -174,6 +176,20 @@ var _ = ginkgo.Describe("[sig-node] ConfigMap", func() {
 		}, metav1.CreateOptions{})
 		framework.ExpectNoError(err, "failed to create ConfigMap")
 
+		ginkgo.By("setting a watch for the ConfigMap")
+		// setup a watch for the ConfigMap
+		resourceWatchTimeoutSeconds := int64(60)
```
Comment (spiffxp) [inline test/e2e/common/configmap.go:181]: Magic number, explain why you are choosing this, preferably extract into a constant up top

File: test/e2e/common/configmap.go [hunk 2]
```diff
@@ -212,10 +236,17 @@ var _ = ginkgo.Describe("[sig-node] ConfigMap", func() {
...
 		err = f.ClientSet.CoreV1().ConfigMaps(testNamespaceName).DeleteCollection(context.TODO(), metav1.DeleteOptions{}, metav1.ListOptions{
 			LabelSelector: "test-configmap-static=true",
 		})
 		framework.ExpectNoError(err, "failed to delete ConfigMap collection with LabelSelector")
+		ginkgo.By("waiting for the ConfigMap to be deleted")
+		for watchEvent := range resourceWatchChan {
+			if watchEvent.Type == watch.Deleted {
+				break
+			}
+		}
```
Comment (spiffxp) [inline test/e2e/common/configmap.go:249]: I think this applies for all of the watch-based checks, but it's sticking out to me here.  Let's say the ConfigMap isn't deleted within 60 seconds, how will this test catch the failure?

In other words, what does the failure of this test look like if the watch times out?

**Review by spiffxp — COMMENTED** [review]
Two nits

File: test/e2e/common/configmap.go (hunk 1 above)
Comment (BobyMCbobs) [reply test/e2e/common/configmap.go]: Adjusted in 387c686bfe1

File: test/e2e/common/configmap.go (hunk 2 above)
Comment (BobyMCbobs) [reply test/e2e/common/configmap.go:254]: I'm not sure that anything will happen on watch timeout. I tried with a timeout of both `0` and `1` in different local tests and this test still succeeded.

Anyhow, I'm not sure how to implement handling of failure (but would also like that to be in there) - are there any places in the test folder where you'd possibly please be able to provide me with an example of how to handle failure on resource deletion?

Comment (hh) [reply test/e2e/common/configmap.go:254]: I also asked in #k8s-conformance slack https://kubernetes.slack.com/archives/C78F00H99/p1585767344008100

**Comment by spiffxp** [comment]
/hold
to discuss failure mode of this test when the watch times out during tomorrow's conformance meeting

**Comment by hh** [comment]
I don't see any flakes here, and I think we should be ok for an /lgtm

**Comment by hh** [comment]
@mtaufen can you take a quick look?

**Comment by hh** [comment]
Dumping all watch events into an array and ensure what you are looking for is received.

File: test/e2e/common/configmap.go (hunk 2 above)
Comment (spiffxp) [reply test/e2e/common/configmap.go:254]: At a minimum this should have a flag set to false before the loop, update the flag to true on the inside of the loop, and do a framework.Expect(...) after the loop, eg:
```
foundWatchDeleteEvent := false
for ... {
 if ... {
  foundWatchDeleteEvent := true
  break
  }
}
framework.ExpectEqual(true, foundWatchDeleteEvent, "expected to find a watch.Delete event configmap %s", configMap)
```

File: test/e2e/common/configmap.go [hunk 3]
```diff
@@ -174,6 +181,19 @@ var _ = ginkgo.Describe("[sig-node] ConfigMap", func() {
...
+		resourceWatch, err := f.ClientSet.CoreV1().ConfigMaps(testNamespaceName).Watch(context.TODO(), metav1.ListOptions{LabelSelector: "test-configmap-static=true", TimeoutSeconds: &generalWatchTimeout})
+		framework.ExpectNoError(err, "Failed to setup watch on newly created ConfigMap")
+
+		resourceWatchChan := resourceWatch.ResultChan()
+		ginkgo.By("waiting for the ConfigMap to be added")
+		for watchEvent := range resourceWatchChan {
+			if watchEvent.Type == watch.Added {
+				break
+			}
+		}
```
Comment (spiffxp) [inline test/e2e/common/configmap.go:195]: At a minimum this should have a flag set to false before the loop, update the flag to true on the inside of the loop, and do a framework.Expect(...) after the loop, eg:
```
foundWatchDeleteEvent := false
for ... {
 if ... {
  foundWatchDeleteEvent := true
  break
  }
}
framework.ExpectEqual(true, foundWatchDeleteEvent, "expected to find a watch.Delete event configmap %s", configMap)
```

File: test/e2e/common/configmap.go (hunk 2 above)
Comment (BobyMCbobs) [reply test/e2e/common/configmap.go:261]: Added in 7ba8b35b9c4

File: test/e2e/common/configmap.go (hunk 3 above)
Comment (BobyMCbobs) [reply test/e2e/common/configmap.go]: Added in 7ba8b35b9c4

**Comment by Riaankl** [comment]
Closing in favour of: #90944 to be placed in PR Needing review column on org dashboard
/close

**Pull request closed without merging**